tests/MongoDB.Bson.Tests/Serialization/Serializers/GuidSerializerTests.cs

```diff
@@ -0,0 +1,612 @@
+        {
+            var data = new TheoryData<GuidRepresentationMode, GuidRepresentation, GuidRepresentation, GuidRepresentation, GuidRepresentation>();
+
+            foreach (var defaultGuidRepresentationMode in EnumHelper.GetValues<GuidRepresentationMode>())
+            {
+                foreach (var defaultGuidRepresentation in EnumHelper.GetValues<GuidRepresentation>())
+                {
+                    if (defaultGuidRepresentationMode == GuidRepresentationMode.V3 && defaultGuidRepresentation != GuidRepresentation.Unspecified)
+                    {
+                        continue;
+                    }
+
+                    foreach (var serializerGuidRepresentation in EnumHelper.GetValues<GuidRepresentation>())
+                    {
+                        if (defaultGuidRepresentationMode == GuidRepresentationMode.V3 && serializerGuidRepresentation == GuidRepresentation.Unspecified)
+                        {
+                            continue;
+                        }
+
+                        foreach (var readerGuidRepresentation in EnumHelper.GetValues<GuidRepresentation>())
+                        {
+                            if (defaultGuidRepresentationMode == GuidRepresentationMode.V2 && 
+                                serializerGuidRepresentation != GuidRepresentation.Unspecified &&
+                                readerGuidRepresentation != GuidRepresentation.Unspecified &&
+                                GuidConverter.GetSubType(serializerGuidRepresentation) != GuidConverter.GetSubType(readerGuidRepresentation))
+                            {
+                                continue;
+                            }
+                            if (defaultGuidRepresentationMode == GuidRepresentationMode.V3 && readerGuidRepresentation != GuidRepresentation.Unspecified)
+                            {
+                                continue;
+                            }
+
+                            var expectedGuidRepresentation = serializerGuidRepresentation;
+                            if (defaultGuidRepresentationMode == GuidRepresentationMode.V2 && expectedGuidRepresentation == GuidRepresentation.Unspecified)
+                            {
+                                expectedGuidRepresentation = readerGuidRepresentation;
+                            }
+                            if (expectedGuidRepresentation == GuidRepresentation.Unspecified)
+                            {
+                                continue;
+                            }
+
+                            data.Add(defaultGuidRepresentationMode, defaultGuidRepresentation, serializerGuidRepresentation, readerGuidRepresentation, expectedGuidRepresentation);
+                        }
+                    }
+                }
+            }
+
+            return data;
+        }
+
+        [Theory]
+        [MemberData(nameof(Deserialize_should_return_expected_result_when_representation_is_binary_MemberData))]
```

Comment:
I was having a difficult time coming with they very large number of combinations that needed to be tested, so instead I wrote a function to generate the desired combinations.

Same for other similar test methods.